From patchwork Wed Oct 10 14:48:50 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: "Emilio G. Cota" <cota@braap.org>
X-Patchwork-Id: 10634695
Return-Path: 
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CE84769B1
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:52:36 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 802D52A50D
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:52:36 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 744822A553; Wed, 10 Oct 2018 14:52:36 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from lists.gnu.org (lists.gnu.org [208.118.235.17])
	(using TLSv1 with cipher AES256-SHA (256/256 bits))
	(No client certificate requested)
	by mail.wl.linuxfoundation.org (Postfix) with ESMTPS id 109282A50D
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:52:35 +0000 (UTC)
Received: from localhost ([::1]:57711 helo=lists.gnu.org)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>)
	id 1gAFql-0005R2-94
	for patchwork-qemu-devel@patchwork.kernel.org;
 Wed, 10 Oct 2018 10:52:35 -0400
Received: from eggs.gnu.org ([2001:4830:134:3::10]:58915)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnU-0002nu-GB
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:13 -0400
Received: from Debian-exim by eggs.gnu.org with spam-scanned (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnP-0004ba-MU
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:12 -0400
Received: from out3-smtp.messagingengine.com ([66.111.4.27]:41295)
	by eggs.gnu.org with esmtps (TLS1.0:DHE_RSA_AES_256_CBC_SHA1:32)
	(Exim 4.71) (envelope-from <cota@braap.org>) id 1gAFnP-0004aP-F0
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:07 -0400
Received: from compute4.internal (compute4.nyi.internal [10.202.2.44])
	by mailout.nyi.internal (Postfix) with ESMTP id 0318C21F86;
	Wed, 10 Oct 2018 10:49:05 -0400 (EDT)
Received: from mailfrontend2 ([10.202.2.163])
	by compute4.internal (MEProxy); Wed, 10 Oct 2018 10:49:05 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=braap.org; h=
	from:to:cc:subject:date:message-id:in-reply-to:references; s=
	mesmtp; bh=kQovQBwyOE8NRXGBmyghO6a5DpCuC7KJZuJpXboCGBw=; b=wGFp7
	1dd1YxNBoBnbGvujoUGiwCn7m4fVPvbvt4ZPXhjNbuHtGLC6pKh4M9M+VmVCiV3o
	d1p8LnNfqfWhHK7B33+aov39KrzdhaglnC3tys5unYy4jZZHuV3jNQ8s+FaM5gAv
	9J3twmtt3kxrVookwnvqyD8Zt8AKYUebQh4B9g=
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
	messagingengine.com; h=cc:date:from:in-reply-to:message-id
	:references:subject:to:x-me-proxy:x-me-proxy:x-me-sender
	:x-me-sender:x-sasl-enc; s=fm1; bh=kQovQBwyOE8NRXGBmyghO6a5DpCuC
	7KJZuJpXboCGBw=; b=xwTKg13WzMfp0NAkk1p+fnJUbVgbWIBlperfCcVuGBGsJ
	m6JcHtAi+wgonz69pseN+MVjE9Vqj895iXYwSVTRnK/8L65MhI6G58jzAQIxgHLQ
	eTWUkNnGVOqYUWrXAdPnt4dEmfGEJ6RQmxBZMTbcC/r05Ze0PSelWlL4mybx9oL2
	lS38YnfkF5FpGEObBKfpipUmSWWAKyXMav0PTpcftaYkbJRsKnVzzni2D4OVBRzG
	Cf5FeC+N8wHiieuomAJLTJL4GDIXhWe7pqgkMyMzcRZsCTZ9BHsyRcLQInNEm7um
	lrmV4A9MelKbl/O8WAS+yUwgXjYHKnKjKDYeUOyRA==
X-ME-Sender: <xms:YBG-W8NkeQkjwgice_uGB7cbAFvwhJOJFIcq7fxKkEKUFGgQK5eToQ>
X-ME-Proxy: <xmx:YBG-WzwiBfofHdzM-6E4bD7P0pQEdxr7nxbuqA0e0h_PchB-BsDw6g>
	<xmx:YBG-W-LdzJYNVtGyvGiSy5s7j-xERi52jMUtssI6MZALxLfA81jGmQ>
	<xmx:YBG-W-IkX65So9wA47gNqh9MTwq9A7uWk2OUvF3XmAoXvhU0KB3BTQ>
	<xmx:YBG-W_WVuuftsrnALl7l2bDIyVk3YZaZp112D6QeEQHbRfX3yKGJ7A>
	<xmx:YBG-Wy8q9Z2mx_DCI9GRCUVtAFscg8SP-FwCu0ShzBssnzY7svhQZw>
	<xmx:YBG-W4V_q0yWAKnO_whZUepJ1MzzyzmlbxGamsXR3qzQTruh5TNFsg>
Received: from localhost (flamenco.cs.columbia.edu [128.59.20.216])
	by mail.messagingengine.com (Postfix) with ESMTPA id 93FA4102F0;
	Wed, 10 Oct 2018 10:49:04 -0400 (EDT)
From: "Emilio G. Cota" <cota@braap.org>
To: qemu-devel@nongnu.org
Date: Wed, 10 Oct 2018 10:48:50 -0400
Message-Id: <20181010144853.13005-2-cota@braap.org>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <20181010144853.13005-1-cota@braap.org>
References: <20181010144853.13005-1-cota@braap.org>
X-detected-operating-system: by eggs.gnu.org: GNU/Linux 2.2.x-3.x [generic]
X-Received-From: 66.111.4.27
Subject: [Qemu-devel] [PATCH 1/4] tcg: access cpu->icount_decr.u16.high with
 atomics
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.21
Precedence: list
List-Id: <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <http://lists.nongnu.org/archive/html/qemu-devel/>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Cc: Richard Henderson <richard.henderson@linaro.org>
Errors-To: 
 qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org
Sender: "Qemu-devel"
	<qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
X-Virus-Scanned: ClamAV using ClamSMTP

Consistently access u16.high with atomics to avoid
undefined behaviour in MTTCG.

Note that icount_decr.u16.low is only used in icount mode,
so regular accesses to it are OK.

Reviewed-by: Richard Henderson <richard.henderson@linaro.org>
Signed-off-by: Emilio G. Cota <cota@braap.org>
---
 accel/tcg/tcg-all.c       | 2 +-
 accel/tcg/translate-all.c | 2 +-
 qom/cpu.c                 | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/accel/tcg/tcg-all.c b/accel/tcg/tcg-all.c
index 56dbb56a16..3d25bdcc17 100644
--- a/accel/tcg/tcg-all.c
+++ b/accel/tcg/tcg-all.c
@@ -51,7 +51,7 @@ static void tcg_handle_interrupt(CPUState *cpu, int mask)
     if (!qemu_cpu_is_self(cpu)) {
         qemu_cpu_kick(cpu);
     } else {
-        cpu->icount_decr.u16.high = -1;
+        atomic_set(&cpu->icount_decr.u16.high, -1);
         if (use_icount &&
             !cpu->can_do_io
             && (mask & ~old_mask) != 0) {
diff --git a/accel/tcg/translate-all.c b/accel/tcg/translate-all.c
index ad5c758246..356dcd0948 100644
--- a/accel/tcg/translate-all.c
+++ b/accel/tcg/translate-all.c
@@ -2341,7 +2341,7 @@ void cpu_interrupt(CPUState *cpu, int mask)
 {
     g_assert(qemu_mutex_iothread_locked());
     cpu->interrupt_request |= mask;
-    cpu->icount_decr.u16.high = -1;
+    atomic_set(&cpu->icount_decr.u16.high, -1);
 }
 
 /*
diff --git a/qom/cpu.c b/qom/cpu.c
index 92599f3541..20ad54d43f 100644
--- a/qom/cpu.c
+++ b/qom/cpu.c
@@ -266,7 +266,7 @@ static void cpu_common_reset(CPUState *cpu)
     cpu->mem_io_pc = 0;
     cpu->mem_io_vaddr = 0;
     cpu->icount_extra = 0;
-    cpu->icount_decr.u32 = 0;
+    atomic_set(&cpu->icount_decr.u32, 0);
     cpu->can_do_io = 1;
     cpu->exception_index = -1;
     cpu->crash_occurred = false;

From patchwork Wed Oct 10 14:48:51 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: "Emilio G. Cota" <cota@braap.org>
X-Patchwork-Id: 10634681
Return-Path: 
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id DFC2746E4
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:50:53 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 9488F2A553
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:50:53 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 88D582A57B; Wed, 10 Oct 2018 14:50:53 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from lists.gnu.org (lists.gnu.org [208.118.235.17])
	(using TLSv1 with cipher AES256-SHA (256/256 bits))
	(No client certificate requested)
	by mail.wl.linuxfoundation.org (Postfix) with ESMTPS id E14382A587
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:50:52 +0000 (UTC)
Received: from localhost ([::1]:57701 helo=lists.gnu.org)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>)
	id 1gAFp4-00044p-DB
	for patchwork-qemu-devel@patchwork.kernel.org;
 Wed, 10 Oct 2018 10:50:50 -0400
Received: from eggs.gnu.org ([2001:4830:134:3::10]:58918)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnU-0002nw-MF
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:13 -0400
Received: from Debian-exim by eggs.gnu.org with spam-scanned (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnP-0004bQ-Lv
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:12 -0400
Received: from out3-smtp.messagingengine.com ([66.111.4.27]:53327)
	by eggs.gnu.org with esmtps (TLS1.0:DHE_RSA_AES_256_CBC_SHA1:32)
	(Exim 4.71) (envelope-from <cota@braap.org>) id 1gAFnP-0004aR-Fp
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:07 -0400
Received: from compute4.internal (compute4.nyi.internal [10.202.2.44])
	by mailout.nyi.internal (Postfix) with ESMTP id 4FA3721D19;
	Wed, 10 Oct 2018 10:49:05 -0400 (EDT)
Received: from mailfrontend2 ([10.202.2.163])
	by compute4.internal (MEProxy); Wed, 10 Oct 2018 10:49:05 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=braap.org; h=
	from:to:cc:subject:date:message-id:in-reply-to:references; s=
	mesmtp; bh=m4aJvnAPBbEyHlFUNekIB6aLbm94eCE+qBmYPRVUKdA=; b=CYJGh
	8T81kWGy7u5moC8/NdtRYQFAmYhf5Yfk9PdZQlLL0yhjRzHeKoKor1BY0JuejhK4
	vbk/lYqBFWAUbupsx21k+q/7cuVxNRXggaZhsD2vm3V6D7w8cKeOmE4WZldpqMiO
	xW8Tgko/ruw4KtRLgyiVDXqDCPA0B3m6wNJROU=
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
	messagingengine.com; h=cc:date:from:in-reply-to:message-id
	:references:subject:to:x-me-proxy:x-me-proxy:x-me-sender
	:x-me-sender:x-sasl-enc; s=fm1; bh=m4aJvnAPBbEyHlFUNekIB6aLbm94e
	CE+qBmYPRVUKdA=; b=PEQm7aPEQHt2x5fIyUhCB6c1EoMOhi9ZAXoIIyNoFZOXD
	yfGjAjFsBPqata68nsa2OUmzYDxSiYNg7tyACFN1VlLuLjxVCurxslPLSIY/3qnS
	QILcbX+glT/PxlEGOoR1pKc4dzFFscEAb48W4EcjFJ8lQsh2pUtQJr2ck7PQ7myy
	QwbYRFV/8MVmlRvLdwMzEZEPYAl0/tkxd9fWbO/3/MmOqRLWsjEFMfZQDEkYeOVX
	HNdVZMN2+wCZs+8bbyDBHmWF6XmVOurHZfYYNvTBAOR+mnd+tbleg/CFHOq6+1L5
	mtUEl5hIEML/THZZpykh0DvlRKy0tKJVxOxKTYxXQ==
X-ME-Sender: <xms:YRG-Wy_euZTTTt3z1H1hGSb_v3J7q90xj0Zh4ncLcR45mRgacAXp1A>
X-ME-Proxy: <xmx:YRG-W0dgRhBG-x4lggKEbNOgN_1PFW_a9eLFrGgseJwh2BZPH6kL6g>
	<xmx:YRG-WyrHHbtGjiXHoMvn6BE1NMmtMuUotSv4EvL-8ndd_j0wL-3QEQ>
	<xmx:YRG-W9QQ_CvOqRL1uaVlqDQ-w2ykAd1r5vPnS8RMFLLTF_D8UpDdww>
	<xmx:YRG-W638Jtk0bMDPeVh0jRhJpofqoLLbolu4zioqgQnnh7qaeEMRbw>
	<xmx:YRG-W4CaGKLmIXVJkcpxEy9g_yEzWx0Sj51Eg8CAFEvCTxNxhev7SQ>
	<xmx:YRG-W798om2nwFmQbvKGGOlj50K_jPjcOcF7nvOUyIwFo-ghudWLHg>
Received: from localhost (flamenco.cs.columbia.edu [128.59.20.216])
	by mail.messagingengine.com (Postfix) with ESMTPA id BD9ED102F2;
	Wed, 10 Oct 2018 10:49:04 -0400 (EDT)
From: "Emilio G. Cota" <cota@braap.org>
To: qemu-devel@nongnu.org
Date: Wed, 10 Oct 2018 10:48:51 -0400
Message-Id: <20181010144853.13005-3-cota@braap.org>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <20181010144853.13005-1-cota@braap.org>
References: <20181010144853.13005-1-cota@braap.org>
X-detected-operating-system: by eggs.gnu.org: GNU/Linux 2.2.x-3.x [generic]
X-Received-From: 66.111.4.27
Subject: [Qemu-devel] [PATCH 2/4] tcg: fix use of uninitialized variable
 under CONFIG_PROFILER
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.21
Precedence: list
List-Id: <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <http://lists.nongnu.org/archive/html/qemu-devel/>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Cc: Richard Henderson <richard.henderson@linaro.org>
Errors-To: 
 qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org
Sender: "Qemu-devel"
	<qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
X-Virus-Scanned: ClamAV using ClamSMTP

We forgot to initialize n in commit 15fa08f845 ("tcg: Dynamically
allocate TCGOps", 2017-12-29).

Signed-off-by: Emilio G. Cota <cota@braap.org>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
---
 tcg/tcg.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tcg/tcg.c b/tcg/tcg.c
index f27b22bd3c..8f26916b99 100644
--- a/tcg/tcg.c
+++ b/tcg/tcg.c
@@ -3430,7 +3430,7 @@ int tcg_gen_code(TCGContext *s, TranslationBlock *tb)
 
 #ifdef CONFIG_PROFILER
     {
-        int n;
+        int n = 0;
 
         QTAILQ_FOREACH(op, &s->ops, link) {
             n++;

From patchwork Wed Oct 10 14:48:52 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: "Emilio G. Cota" <cota@braap.org>
X-Patchwork-Id: 10634701
Return-Path: 
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8BA1D46E4
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:54:57 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6E1CA2A3D5
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:54:57 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 621892A410; Wed, 10 Oct 2018 14:54:57 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from lists.gnu.org (lists.gnu.org [208.118.235.17])
	(using TLSv1 with cipher AES256-SHA (256/256 bits))
	(No client certificate requested)
	by mail.wl.linuxfoundation.org (Postfix) with ESMTPS id C5DBC2A3D5
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:54:56 +0000 (UTC)
Received: from localhost ([::1]:57720 helo=lists.gnu.org)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>)
	id 1gAFt1-0007Cr-GH
	for patchwork-qemu-devel@patchwork.kernel.org;
 Wed, 10 Oct 2018 10:54:55 -0400
Received: from eggs.gnu.org ([2001:4830:134:3::10]:58921)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnU-0002nx-Mj
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:13 -0400
Received: from Debian-exim by eggs.gnu.org with spam-scanned (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnP-0004bf-MV
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:12 -0400
Received: from out3-smtp.messagingengine.com ([66.111.4.27]:48271)
	by eggs.gnu.org with esmtps (TLS1.0:DHE_RSA_AES_256_CBC_SHA1:32)
	(Exim 4.71) (envelope-from <cota@braap.org>) id 1gAFnP-0004aZ-F1
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:07 -0400
Received: from compute4.internal (compute4.nyi.internal [10.202.2.44])
	by mailout.nyi.internal (Postfix) with ESMTP id 5B2C822123;
	Wed, 10 Oct 2018 10:49:05 -0400 (EDT)
Received: from mailfrontend2 ([10.202.2.163])
	by compute4.internal (MEProxy); Wed, 10 Oct 2018 10:49:05 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=braap.org; h=
	from:to:cc:subject:date:message-id:in-reply-to:references; s=
	mesmtp; bh=RnnyextfCFUjAOna+t7W8gWY39U5DeplciMPVmW8uE0=; b=bQBVe
	i694fQxDiaixsyQu76vSNIi8vWgjcUEuSWz5O/uOGZXVNLtaV6hfyOuMz35uOAEp
	roHRkZ6sNoLDgsR3hP9GLZmNYqdDzOgvQyVi7uvmrjAZj0iorSuwfcBN15Zzs6XT
	DCZv7SCGpg8ohlLw/cpsiSesAuG6swDqsXBqLw=
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
	messagingengine.com; h=cc:date:from:in-reply-to:message-id
	:references:subject:to:x-me-proxy:x-me-proxy:x-me-sender
	:x-me-sender:x-sasl-enc; s=fm1; bh=RnnyextfCFUjAOna+t7W8gWY39U5D
	eplciMPVmW8uE0=; b=JEk/25yky5dJT7OH4bhymj46EIdgvB6xGctlb/mP5RNhc
	EXJNX/DKlpZlNC4ClRFohIhc9oZk7H9SE7csfTZ8wghuYunv4sCN/My8prKGfPC5
	Gw73TF203tywiRWPQZvB8SiaOMGiZ8L1A6sJPsGdMZPhmj89KgbSXf/m4mAFyyKo
	7nF8FBw6b7zaYTdnc0bBmETU+/3G5Tp44qYsCcQO0yfcAea2jTmCu5b4gUOb9Xtr
	5DYsZD0XznaCVepTvuY0/rHjTYG15UelTW9dUbVJC6oUo2WYqxbwfhxQM+GczgeL
	NBiBZalnH8+W7CcA8BCfENSh4Kn2ol831Rv8Aiplw==
X-ME-Sender: <xms:YRG-W0BDxl-doyTLZ-wnaQZIdlP6tOOFaXf3maT_r9-jS_vJTKkMTA>
X-ME-Proxy: <xmx:YRG-W-Bqa4u-MSkKXZDGS2JgUOIrb4FSr6qmvVbTtRnFDRepCzw-sQ>
	<xmx:YRG-W_MRwlSLBLLxcSaI-GEZRvPrEyWPEoBDGFGKoD-vbKFrbAobVA>
	<xmx:YRG-W7zEkToZNTwz5yZ0Tjnio9LrZujOZKLXp3e52HI2DFEiMNeVwg>
	<xmx:YRG-W9bJJmMT3GFBjxQXy-Pz3guj8UPHg6XPwipAnr624WC3r0gIBQ>
	<xmx:YRG-W3C9-4l3x0-bFq5FHBS-IhsAfHqCxkqCaJaGRJ2j6gPidiMI2w>
	<xmx:YRG-W-zvM8vUwQxdDgiNWy1Q5UbHjlqJ-4jVwGqrfToGhd7ozaygMg>
Received: from localhost (flamenco.cs.columbia.edu [128.59.20.216])
	by mail.messagingengine.com (Postfix) with ESMTPA id F1DC5102EC;
	Wed, 10 Oct 2018 10:49:04 -0400 (EDT)
From: "Emilio G. Cota" <cota@braap.org>
To: qemu-devel@nongnu.org
Date: Wed, 10 Oct 2018 10:48:52 -0400
Message-Id: <20181010144853.13005-4-cota@braap.org>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <20181010144853.13005-1-cota@braap.org>
References: <20181010144853.13005-1-cota@braap.org>
X-detected-operating-system: by eggs.gnu.org: GNU/Linux 2.2.x-3.x [generic]
X-Received-From: 66.111.4.27
Subject: [Qemu-devel] [PATCH 3/4] tcg: plug holes in struct TCGProfile
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.21
Precedence: list
List-Id: <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <http://lists.nongnu.org/archive/html/qemu-devel/>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Cc: Richard Henderson <richard.henderson@linaro.org>
Errors-To: 
 qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org
Sender: "Qemu-devel"
	<qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
X-Virus-Scanned: ClamAV using ClamSMTP

This plugs two 4-byte holes in 64-bit.

Signed-off-by: Emilio G. Cota <cota@braap.org>
---
 tcg/tcg.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tcg/tcg.h b/tcg/tcg.h
index f9f12378e9..d80ef2a883 100644
--- a/tcg/tcg.h
+++ b/tcg/tcg.h
@@ -633,8 +633,8 @@ typedef struct TCGProfile {
     int64_t tb_count;
     int64_t op_count; /* total insn count */
     int op_count_max; /* max insn per TB */
-    int64_t temp_count;
     int temp_count_max;
+    int64_t temp_count;
     int64_t del_op_count;
     int64_t code_in_len;
     int64_t code_out_len;

From patchwork Wed Oct 10 14:48:53 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: "Emilio G. Cota" <cota@braap.org>
X-Patchwork-Id: 10634679
Return-Path: 
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7E56269B1
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:50:55 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 665312A565
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:50:55 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 64DF22A56B; Wed, 10 Oct 2018 14:50:55 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from lists.gnu.org (lists.gnu.org [208.118.235.17])
	(using TLSv1 with cipher AES256-SHA (256/256 bits))
	(No client certificate requested)
	by mail.wl.linuxfoundation.org (Postfix) with ESMTPS id D29492A586
	for <patchwork-qemu-devel@patchwork.kernel.org>;
 Wed, 10 Oct 2018 14:50:54 +0000 (UTC)
Received: from localhost ([::1]:57703 helo=lists.gnu.org)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from
 <qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>)
	id 1gAFp8-00047E-4A
	for patchwork-qemu-devel@patchwork.kernel.org;
 Wed, 10 Oct 2018 10:50:54 -0400
Received: from eggs.gnu.org ([2001:4830:134:3::10]:58917)
	by lists.gnu.org with esmtp (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnU-0002nv-MH
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:14 -0400
Received: from Debian-exim by eggs.gnu.org with spam-scanned (Exim 4.71)
	(envelope-from <cota@braap.org>) id 1gAFnP-0004bc-Mk
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:12 -0400
Received: from out3-smtp.messagingengine.com ([66.111.4.27]:57409)
	by eggs.gnu.org with esmtps (TLS1.0:DHE_RSA_AES_256_CBC_SHA1:32)
	(Exim 4.71) (envelope-from <cota@braap.org>) id 1gAFnP-0004ah-F1
	for qemu-devel@nongnu.org; Wed, 10 Oct 2018 10:49:07 -0400
Received: from compute4.internal (compute4.nyi.internal [10.202.2.44])
	by mailout.nyi.internal (Postfix) with ESMTP id 8AEDF21F03;
	Wed, 10 Oct 2018 10:49:05 -0400 (EDT)
Received: from mailfrontend2 ([10.202.2.163])
	by compute4.internal (MEProxy); Wed, 10 Oct 2018 10:49:05 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=braap.org; h=
	from:to:cc:subject:date:message-id:in-reply-to:references; s=
	mesmtp; bh=+wZF9DPuJOSNNhPKQUg8G+MTdjeU3DkiGqMf92cqlDg=; b=UWF8c
	fUiUIKOxFrd/88On2nt9zaFYWlIs/d8SWl1KCL0nwgqx1xQctZ15r0slyEZ1e0j+
	weLsvvO76RrU4cHupYs8bhtHSKMNZhDP65tttKQkwNuaast7WM0Rke7iLiyscZPY
	G+8whta5sToIaZgIrNIeuHAvdviCgJsZNgnXwg=
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
	messagingengine.com; h=cc:date:from:in-reply-to:message-id
	:references:subject:to:x-me-proxy:x-me-proxy:x-me-sender
	:x-me-sender:x-sasl-enc; s=fm1; bh=+wZF9DPuJOSNNhPKQUg8G+MTdjeU3
	DkiGqMf92cqlDg=; b=CwmSy46qzukBzvPl53zdAmoWU1gjSfmAQC1y/PQKT8Kyw
	i+UzZR+vRW8CSAEE3IyenDHZDt9aAsWz2C1fDzQfFuZxh5KutNwiqrS2/r72uoi0
	tQWX/FACwnAg/e5hOoSh8MiIMZvRjublJNNM0i6xS2r/plOrKaEstPXZO3tfwqje
	LAHoCf9EJpY7Mlu3/ki5KJJL07igTmbOtzUJ4fs1C3ifpD1hIKFaV4ma0O34SCaU
	631SC3CoyhIrKe2Qq5YFF7+ixN66fKfca0z+Q+iM+2JmFUG2NZoyUnaR1ixJrFIl
	JQhd4MYafbP70SNGZhC0+Z8mGbqEx5mlWhvRJtGtA==
X-ME-Sender: <xms:YRG-W0ECRPfGGSHy1m8M4BjO0VoRxGAgGTd6y9KKZ8xl8UUu6L-SQg>
X-ME-Proxy: <xmx:YRG-W35qyNKgV7AvK6KDTKE3ruI_IY5DK-A2MjGXmUQjObjr-msraQ>
	<xmx:YRG-Wzd5Uc-st6Ig3UTeuDGbBkktKmMz1dLD_EE4CCZqv4O-m00zaw>
	<xmx:YRG-W7ObT9lglHTHb4s2AQn5n34q5Rm9vDYWK9qp_dhc_96T49Pfmg>
	<xmx:YRG-WwwIL3I7zI_PPlNNQ29uqdHRsJIVMFwCFmpojsm9uRj70Z9vlg>
	<xmx:YRG-W_760kON2yLTuiIVS8i9vhYMlA68-Xt_kmKZMHXrHi0jLFQQ0g>
	<xmx:YRG-W844_upSWDlbk_uuRkSzqav9qLfBGuH1jnL3Oc-Onlws2p1bHg>
Received: from localhost (flamenco.cs.columbia.edu [128.59.20.216])
	by mail.messagingengine.com (Postfix) with ESMTPA id 30B31102A0;
	Wed, 10 Oct 2018 10:49:05 -0400 (EDT)
From: "Emilio G. Cota" <cota@braap.org>
To: qemu-devel@nongnu.org
Date: Wed, 10 Oct 2018 10:48:53 -0400
Message-Id: <20181010144853.13005-5-cota@braap.org>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <20181010144853.13005-1-cota@braap.org>
References: <20181010144853.13005-1-cota@braap.org>
X-detected-operating-system: by eggs.gnu.org: GNU/Linux 2.2.x-3.x [generic]
X-Received-From: 66.111.4.27
Subject: [Qemu-devel] [PATCH 4/4] tcg: distribute tcg_time into TCG contexts
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.21
Precedence: list
List-Id: <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <http://lists.nongnu.org/archive/html/qemu-devel/>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
	<mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Cc: Richard Henderson <richard.henderson@linaro.org>
Errors-To: 
 qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org
Sender: "Qemu-devel"
	<qemu-devel-bounces+patchwork-qemu-devel=patchwork.kernel.org@nongnu.org>
X-Virus-Scanned: ClamAV using ClamSMTP

When we implemented per-vCPU TCG contexts, we forgot to also
distribute the tcg_time counter, which has remained as a global
accessed without any serialization, leading to potentially missed
counts.

Fix it by distributing the field over the TCG contexts, embedding
it into TCGProfile with a field called "cpu_exec_time", which is more
descriptive than "tcg_time". Add a function to query this value
directly, and for completeness, fill in the field in
tcg_profile_snapshot, even though its callers do not use it.

Signed-off-by: Emilio G. Cota <cota@braap.org>
---
 include/qemu/timer.h |  1 -
 tcg/tcg.h            |  2 ++
 cpus.c               |  3 ++-
 monitor.c            | 13 ++++++++++---
 tcg/tcg.c            | 23 +++++++++++++++++++++++
 5 files changed, 37 insertions(+), 5 deletions(-)

diff --git a/include/qemu/timer.h b/include/qemu/timer.h
index a005ed2692..dfecd03e28 100644
--- a/include/qemu/timer.h
+++ b/include/qemu/timer.h
@@ -1046,7 +1046,6 @@ static inline int64_t profile_getclock(void)
     return get_clock();
 }
 
-extern int64_t tcg_time;
 extern int64_t dev_time;
 #endif
 
diff --git a/tcg/tcg.h b/tcg/tcg.h
index d80ef2a883..c59f254e27 100644
--- a/tcg/tcg.h
+++ b/tcg/tcg.h
@@ -629,6 +629,7 @@ typedef struct TCGOp {
 QEMU_BUILD_BUG_ON(NB_OPS > (1 << 8));
 
 typedef struct TCGProfile {
+    int64_t cpu_exec_time;
     int64_t tb_count1;
     int64_t tb_count;
     int64_t op_count; /* total insn count */
@@ -1002,6 +1003,7 @@ int tcg_check_temp_count(void);
 #define tcg_check_temp_count() 0
 #endif
 
+int64_t tcg_cpu_exec_time(void);
 void tcg_dump_info(FILE *f, fprintf_function cpu_fprintf);
 void tcg_dump_op_count(FILE *f, fprintf_function cpu_fprintf);
 
diff --git a/cpus.c b/cpus.c
index 361678e459..cce64874e6 100644
--- a/cpus.c
+++ b/cpus.c
@@ -1425,7 +1425,8 @@ static int tcg_cpu_exec(CPUState *cpu)
     ret = cpu_exec(cpu);
     cpu_exec_end(cpu);
 #ifdef CONFIG_PROFILER
-    tcg_time += profile_getclock() - ti;
+    atomic_set(&tcg_ctx->prof.cpu_exec_time,
+               tcg_ctx->prof.cpu_exec_time + profile_getclock() - ti);
 #endif
     return ret;
 }
diff --git a/monitor.c b/monitor.c
index b9258a7438..823b5a1099 100644
--- a/monitor.c
+++ b/monitor.c
@@ -83,6 +83,7 @@
 #include "sysemu/cpus.h"
 #include "sysemu/iothread.h"
 #include "qemu/cutils.h"
+#include "tcg/tcg.h"
 
 #if defined(TARGET_S390X)
 #include "hw/s390x/storage-keys.h"
@@ -1966,16 +1967,22 @@ static void hmp_info_numa(Monitor *mon, const QDict *qdict)
 
 #ifdef CONFIG_PROFILER
 
-int64_t tcg_time;
 int64_t dev_time;
 
 static void hmp_info_profile(Monitor *mon, const QDict *qdict)
 {
+    static int64_t last_cpu_exec_time;
+    int64_t cpu_exec_time;
+    int64_t delta;
+
+    cpu_exec_time = tcg_cpu_exec_time();
+    delta = cpu_exec_time - last_cpu_exec_time;
+
     monitor_printf(mon, "async time  %" PRId64 " (%0.3f)\n",
                    dev_time, dev_time / (double)NANOSECONDS_PER_SECOND);
     monitor_printf(mon, "qemu time   %" PRId64 " (%0.3f)\n",
-                   tcg_time, tcg_time / (double)NANOSECONDS_PER_SECOND);
-    tcg_time = 0;
+                   delta, delta / (double)NANOSECONDS_PER_SECOND);
+    last_cpu_exec_time = cpu_exec_time;
     dev_time = 0;
 }
 #else
diff --git a/tcg/tcg.c b/tcg/tcg.c
index 8f26916b99..e85133ef05 100644
--- a/tcg/tcg.c
+++ b/tcg/tcg.c
@@ -30,6 +30,7 @@
 /* Define to jump the ELF file used to communicate with GDB.  */
 #undef DEBUG_JIT
 
+#include "qemu/error-report.h"
 #include "qemu/cutils.h"
 #include "qemu/host-utils.h"
 #include "qemu/timer.h"
@@ -3361,6 +3362,7 @@ void tcg_profile_snapshot(TCGProfile *prof, bool counters, bool table)
         const TCGProfile *orig = &s->prof;
 
         if (counters) {
+            PROF_ADD(prof, orig, cpu_exec_time);
             PROF_ADD(prof, orig, tb_count1);
             PROF_ADD(prof, orig, tb_count);
             PROF_ADD(prof, orig, op_count);
@@ -3412,11 +3414,32 @@ void tcg_dump_op_count(FILE *f, fprintf_function cpu_fprintf)
                     prof.table_op_count[i]);
     }
 }
+
+int64_t tcg_cpu_exec_time(void)
+{
+    unsigned int n_ctxs = atomic_read(&n_tcg_ctxs);
+    unsigned int i;
+    int64_t ret = 0;
+
+    for (i = 0; i < n_ctxs; i++) {
+        const TCGContext *s = atomic_read(&tcg_ctxs[i]);
+        const TCGProfile *prof = &s->prof;
+
+        ret += atomic_read(&prof->cpu_exec_time);
+    }
+    return ret;
+}
 #else
 void tcg_dump_op_count(FILE *f, fprintf_function cpu_fprintf)
 {
     cpu_fprintf(f, "[TCG profiler not compiled]\n");
 }
+
+int64_t tcg_cpu_exec_time(void)
+{
+    error_report("%s: TCG profiler not compiled", __func__);
+    exit(EXIT_FAILURE);
+}
 #endif
 
 
